pytest_mpi/parallel_assert.py

```diff
@@ -0,0 +1,40 @@
+from mpi4py import MPI
+
+def parallel_assert(assertion, subset=None, msg=""):
+    """Make an assertion across MPI.COMM_WORLD
```

Comment:
Is it right to always be COMM_WORLD, should this not take a `comm` argument?

Reply:
Yes, always. The idea here being that you need to fail on all ranks of COMM_WORLD simultaneously.

If this method were to take a comm argument then it would have to determine which ranks of COMM_WORLD were also part of comm and I haven't thought of a good way of doing this, hence currently taking a subset.

Reply:
Ah yes of course. I think `subset` is the right thing to do. "determine which ranks of COMM_WORLD were also part of comm" seems like it might be impossible.

Please make it clear in the docstring why we must have COMM_WORLD.

Reply:
Will do!

Reply:
Do all ranks need to know which other ranks are involved in the assertion? Why not:
```python
def parallel_assert(assertion, participating=True, msg="")
    if participating:
        evaluation = assertion()
    else:
        evaluation = True
```

Reply:
It wouldn't be practical here, but for comparing comms MPI has set operations for processor groups which can be handy.

Reply:
@JHopeCollins won't this fail because we need to know the communicator to do the `allgather`?

Reply:
Isn't the point of this that the comm is always `COMM_WORLD`?

Reply:
Ah I see. Yes your solution is more elegant.

Reply:
We could perhaps merge this with the assertion result.
`parallel_assert(None, ...)`
(Passing `None` for the assertion result) could flag that this rank is not participating.

Reply:
Potentially I guess. That works now but you lose some information so I wonder if it would be restrictive in the future.
But I can't think of a time you would want to pass an assertion result on a rank that wasn't participating, so it's probably fine.

Reply:
See below as to why I don't think `parallel_assert(None, ...)` is a good idea